src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSScorer.java

```diff
@@ -221,16 +225,27 @@ public static void writeMissingReferenceAccessions(final String path, final SAMF
         }
 
         //Find and return all alignments meeting the coverage/identity criteria
-        final Collection<String> hits = new ArrayList<>();
-        if (isValidAlignment(read.getCigar(), read.getAttributeAsInteger("NM"), minCoverage, minIdentity)) {
-            hits.add(SAMSequenceRecord.truncateSequenceName(read.getAssignedContig()));
+        final ArrayList<PSPathogenHitAlignment> hits = new ArrayList<>();
+        final double minIdentityBases = minIdentity * read.getLength();
+        final int numMismatches = read.getAttributeAsInteger("NM");
+        final int numMatches = HostAlignmentReadFilter.getMatchesLessDeletions(read.getCigar(), numMismatches);
+        if (numMatches >= minIdentityBases) {
+            final String recordName = SAMSequenceRecord.truncateSequenceName(read.getAssignedContig());
+            hits.add(new PSPathogenHitAlignment(numMatches, recordName, read.getCigar()));
         }
-        hits.addAll(getValidAlternateHits(read, "XA", 0, 2, 3, minCoverage, minIdentity));
-        hits.addAll(getValidAlternateHits(read, "SA", 0, 3, 5, minCoverage, minIdentity));
+        hits.addAll(getValidAlternateHits(read, "XA", 0, 2, 3, minIdentityBases));
+        hits.addAll(getValidAlternateHits(read, "SA", 0, 3, 5, minIdentityBases));
+
+        if (hits.isEmpty()) return Collections.emptySet();
+
+        //Throw out reads below the identity margin
+        final int maxMatches = hits.stream().mapToInt(PSPathogenHitAlignment::getNumMatches).max().getAsInt();
+        final double minIdentityBasesMargin = (1. - identityMargin) * maxMatches;
+        final List<PSPathogenHitAlignment> bestHits = hits.stream().filter(hit -> hit.getNumMatches() >= minIdentityBasesMargin).collect(Collectors.toList());
 
         //Throw out duplicates and accessions not in the taxonomic database so it returns a list of unique tax ID's
         // for each read in the pair
-        return hits.stream().map(contig -> taxonomyDatabase.accessionToTaxId.containsKey(contig) ? taxonomyDatabase.accessionToTaxId.get(contig) : null)
+        return bestHits.stream().map(hit -> taxonomyDatabase.accessionToTaxId.containsKey(hit.getAccession()) ? taxonomyDatabase.accessionToTaxId.get(hit.getAccession()) : null)
```

Comment:
I would use a filter operator rather than clean the nulls a posteriori.
Otherwise the clean up would be more justified could be justified if you use ```getOrDefault(key, null)``` saving an additional hashmap.get operation. 

If you expect most elements to past the filter then the latter is a better solution in terms of performance, if the oposite the is better to filter and avoid the null clean-up.

Reply:
I expect most will pass. Never knew about getOrDefault() - very handy!